storm-core/src/jvm/org/apache/storm/StormSubmitter.java

```diff
@@ -238,39 +239,24 @@ public static void submitTopologyAs(String name, Map stormConf, StormTopology to
                 }
 
                 // Dependency uploading only makes sense for distributed mode
-                List<String> jarsBlobKeys;
+                List<String> jarsBlobKeys = Collections.emptyList();
                 List<String> artifactsBlobKeys;
 
                 DependencyUploader uploader = new DependencyUploader();
                 try {
                     uploader.init();
                     jarsBlobKeys = uploadDependencyJarsToBlobStore(uploader);
                     artifactsBlobKeys = uploadDependencyArtifactsToBlobStore(uploader);
+                    setDependencyBlobsToTopology(topology, jarsBlobKeys, artifactsBlobKeys);
+                    submitTopologyInDistributeMode(name, topology, opts, progressListener, asUser, conf, serConf);
+                } catch (Throwable e) {
+                    // remove uploaded jars blobs, not artifacts since they're shared across the cluster
+                    uploader.deleteBlobs(jarsBlobKeys);
```

Comment:
@satishd 
Yes I think the scenarios you stated can be occurred. But also Nimbus throws TException when Nimbus is unable to handle the submission, say, runtime exception occurred.

So it's more like loosely catching with less risk vs hardly catching with more risk.
(Risk could be reduced via checking topology name once again indeed.)

After thinking it again, leaving blobs seems less harmful than whole contents of topology couldn't be downloaded at all. I'll follow your suggestion.